client/src/main/java/io/streamnative/oxia/client/grpc/OxiaWriteStreamManager.java

```diff
@@ -34,18 +35,28 @@ public OxiaWriteStreamManager(OxiaStubProvider provider) {
             Metadata.Key.of("shard-id", Metadata.ASCII_STRING_MARSHALLER);
 
     public WriteStreamWrapper getWriteStream(long shardId) {
-        return writeStreams.compute(
-                shardId,
-                (key, stream) -> {
-                    if (stream == null || !stream.isValid()) {
-                        Metadata headers = new Metadata();
-                        headers.put(NAMESPACE_KEY, provider.getNamespace());
-                        headers.put(SHARD_ID_KEY, String.format("%d", shardId));
-                        final var asyncStub = provider.getStubForShard(shardId).async();
-                        return new WriteStreamWrapper(
-                                asyncStub.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers)));
-                    }
-                    return stream;
-                });
+        WriteStreamWrapper wrapper = null;
+        for (int i = 0; i < 2; i++) {
+            wrapper = writeStreams.get(shardId); // lock free first
+            if (wrapper == null) {
+                wrapper =
+                        writeStreams.computeIfAbsent(
+                                shardId,
+                                (__) -> {
+                                    Metadata headers = new Metadata();
+                                    headers.put(NAMESPACE_KEY, provider.getNamespace());
+                                    headers.put(SHARD_ID_KEY, String.format("%d", shardId));
+                                    final var asyncStub = provider.getStubForShard(shardId).async();
+                                    return new WriteStreamWrapper(
+                                            asyncStub.withInterceptors(
+                                                    MetadataUtils.newAttachHeadersInterceptor(headers)));
+                                });
+            }
+            if (wrapper.isValid()) {
+                break;
+            }
+            writeStreams.remove(shardId, wrapper);
+        }
+        return wrapper;
```

Comment:
Will it return null after the second try?

Reply:
Since we have `computIfAbsent`. that should impossible.